Add morse encoding

oliverlee opened the pull request.

oliverlee committed:
Add Mark struct

Add a Mark struct that represents either a dot or dash.

oliverlee committed:
Add Letter struct

Define Morse code for letters A-Z, 0-9, and punctuation marks
%'@)(:,=!.-+"?/.

oliverlee committed:
Implement trait TryFrom for Letter

TryFrom is implemented for '&char' instead of 'char' due to a conflict
with the blanket implementation.

For more details see:
https://github.com/rust-lang/rust/issues/50133
https://old.reddit.com/r/rust/comments/c1yubv/why_this_code_does_not_compile/

oliverlee committed:
Add Word struct

Add a struct that represents a word in Morse Code. Parses from a string.

oliverlee committed:
Fix clippy lint errors

oliverlee committed:
Add Phrase struct

oliverlee committed:
Add timing phrase

Output the timing phrase for a morse code message. This phrase contains
the two states needed for transmission.

mickvangelderen commented on src/phrase.rs:
```diff
@@ -0,0 +1,71 @@
+            "{}",
+            self.words
+                .iter()
+                .map(std::string::ToString::to_string)
```
nice, was going to suggest this further down.

oliverlee committed:
Move coding structs to a code directory

Move structs related to encoding a message to the code directory. Remove
public visibility except for phrase::Phrase and word::ParseWordError as
a user should not need to interact with lower level objects and rename
phrase::Phrase -> Code
word::ParseWordError -> ParseCodeError

oliverlee committed:
Restrict non-public struct visibility

Restrict visibility of {Letter, Mark, Word} to the crate::code module.
